app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,188 @@
+                buildForegroundNotification(getString(R.string.hotspot_running)));
+          }
+          Toast.makeText(this, R.string.server_started__successfully_toast_message,
+              Toast.LENGTH_LONG).show();
+        }
+
+        break;
+
+      case ACTION_STOP_SERVER:
+        stopHotspotAndDismissNotification();
+        break;
+      default:
+        break;
+    }
+    return START_NOT_STICKY;
+  }
+
+  @Nullable @Override public IBinder onBind(@Nullable Intent intent) {
+    return serviceBinder;
+  }
+
+  private Notification buildForegroundNotification(String status) {
+    Log.v(TAG, "Building notification " + status);
+    builder = new NotificationCompat.Builder(this);
+    builder.setContentTitle(getString(R.string.hotspot_notification_content_title))
+        .setContentText(status);
+    Intent targetIntent = new Intent(this, ZimHostActivity.class);
+    targetIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    PendingIntent contentIntent =
+        PendingIntent.getActivity(this, 0, targetIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+    builder.setContentIntent(contentIntent)
+        .setSmallIcon(R.mipmap.kiwix_icon)
+        .setWhen(System.currentTimeMillis());
+
+    hotspotNotificationChannel();
+
+    Intent stopIntent = new Intent(ACTION_STOP);
+    PendingIntent stopHotspot =
+        PendingIntent.getBroadcast(this, 0, stopIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+    builder.addAction(R.drawable.ic_close_white_24dp, getString(R.string.stop_hotspot_button),
+        stopHotspot);
+
+    return (builder.build());
+  }
+
+  //Dismiss notification and turn off hotspot for devices>=O
+  void stopHotspotAndDismissNotification() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      hotspotManager.turnOffHotspot();
+    }
+    webServerHelper.stopAndroidWebServer(serverStateListener);
+    stopForeground(true);
+    stopSelf();
+    notificationManager.cancel(HOTSPOT_NOTIFICATION_ID);
+  }
+
+  @Override
+  public void onDestroy() {
+    if (stopReceiver != null) {
+      unregisterReceiver(stopReceiver);
+    }
+    super.onDestroy();
+  }
+
+  private void hotspotNotificationChannel() {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+      NotificationChannel hotspotServiceChannel = new NotificationChannel(
+          Constants.HOTSPOT_SERVICE_CHANNEL_ID, getString(R.string.hotspot_service_channel_name),
+          NotificationManager.IMPORTANCE_DEFAULT);
+      hotspotServiceChannel.setDescription(getString(R.string.hotspot_channel_description));
+      hotspotServiceChannel.setSound(null, null);
+      builder.setChannelId(Constants.HOTSPOT_SERVICE_CHANNEL_ID);
+      notificationManager.createNotificationChannel(hotspotServiceChannel);
+    }
+  }
+
+  public class HotspotBinder extends Binder {
+
+    @NonNull public HotspotService getService() {
+      return HotspotService.this;
+    }
+  }
+
+  public void registerCallBack(@Nullable ServerStateListener myCallback) {
```

Comment:
can you move this method above the Binder, I totally misread it as being a part of the binder.
place cursor on the method defintion `ctrl+shift+up`

Reply:
Fixed. 6526db70b4cc3fc67cb85ab28b4c7b8ce0a28e02
Thanks for the shortcut. :thumbsup: